File: salesflow-app/src/backend/app/api/routes/live_assist.py
"""
╔════════════════════════════════════════════════════════════════════════════╗
║  LIVE ASSIST API ROUTES                                                    ║
║  Echtzeit-Verkaufsassistenz während Kundengesprächen                       ║
╚════════════════════════════════════════════════════════════════════════════╝

Endpoints:
    
    SESSION MANAGEMENT:
    - POST   /live-assist/start           - Session starten
    - POST   /live-assist/end             - Session beenden
    - GET    /live-assist/session/{id}    - Session-Stats abrufen
    
    QUERY PROCESSING:
    - POST   /live-assist/query           - Live-Anfrage verarbeiten
    
    QUICK ACCESS (ohne Session):
    - GET    /live-assist/facts/{company_id}     - Quick Facts abrufen
    - GET    /live-assist/objections/{company_id} - Einwand-Antworten abrufen
    - GET    /live-assist/knowledge/{vertical}    - Branchenwissen abrufen
"""

from fastapi import APIRouter, Depends, HTTPException, Query, WebSocket, WebSocketDisconnect
from typing import List, Optional, Any
from uuid import UUID

from ...db.deps import get_db, get_current_user, CurrentUser
from ..schemas.live_assist import (
    # Session
    StartSessionRequest,
    StartSessionResponse,
    LiveQueryRequest,
    LiveQueryResponse,
    EndSessionRequest,
    SessionStatsResponse,
    # Quick Access
    AssistIntent,
    ObjectionType,
    FactType,
)
from ...services.live_assist import LiveAssistService
from ...services.live_assist.service_v3 import LiveAssistServiceV3
from ...services.live_assist.coach_analytics import CoachAnalyticsService
from ..schemas.live_assist import (
    # Existing imports already there
    CoachInsightsResponse,
    QueryFeedbackRequest,
    PerformanceMetricsResponse,
    ObjectionAnalyticsItem,
)

router = APIRouter(prefix="/live-assist", tags=["live-assist"])


# =============================================================================
# SESSION MANAGEMENT
# =============================================================================

@router.post("/start", response_model=StartSessionResponse)
async def start_session(
    request: StartSessionRequest,
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Startet eine Live Assist Session.
    
    Trigger: User sagt "Bin mit Kunde" oder aktiviert manuell.
    
    Lädt Key Facts und Produkte vor für schnellen Zugriff.
    """
    service = LiveAssistService(db)
    return service.start_session(str(current_user.id), request)


@router.post("/end")
async def end_session(
    request: EndSessionRequest,
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Beendet eine Live Assist Session.
    
    Optional: Outcome und Feedback angeben.
    """
    service = LiveAssistService(db)
    return service.end_session(str(current_user.id), request)


@router.get("/session/{session_id}", response_model=SessionStatsResponse)
async def get_session_stats(
    session_id: str,
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """Holt Statistiken einer Session."""
    service = LiveAssistService(db)
    return service.get_session_stats(str(current_user.id), session_id)


# =============================================================================
# QUERY PROCESSING
# =============================================================================

@router.post("/query", response_model=LiveQueryResponse)
async def process_query(
    request: LiveQueryRequest,
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Verarbeitet eine Live-Anfrage.
    
    Schnelle Antwort - optimiert für Echtzeit.
    
    Flow:
    1. Intent Detection
    2. Knowledge Retrieval (DB first)
    3. AI Generation (fallback)
    4. Response Logging
    """
    service = LiveAssistService(db)
    return service.process_query(str(current_user.id), request)


# =============================================================================
# QUICK ACCESS (ohne Session)
# =============================================================================

@router.get("/facts/{company_id}")
async def get_quick_facts(
    company_id: str,
    fact_type: Optional[str] = Query(None, description="Filter nach Fact-Typ"),
    key_only: bool = Query(False, description="Nur Key Facts"),
    limit: int = Query(10, ge=1, le=50),
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Holt Quick Facts für eine Firma.
    
    Kann auch ohne aktive Session verwendet werden.
    """
    service = LiveAssistService(db)
    return service.get_quick_facts(
        company_id=company_id,
        fact_type=fact_type,
        key_only=key_only,
        limit=limit,
    )


@router.get("/facts")
async def get_all_quick_facts(
    vertical: Optional[str] = Query(None, description="Filter nach Vertical"),
    fact_type: Optional[str] = Query(None, description="Filter nach Fact-Typ"),
    key_only: bool = Query(False, description="Nur Key Facts"),
    limit: int = Query(10, ge=1, le=50),
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Holt Quick Facts (ohne Company-Filter).
    """
    service = LiveAssistService(db)
    return service.get_quick_facts(
        vertical=vertical,
        fact_type=fact_type,
        key_only=key_only,
        limit=limit,
    )


@router.get("/objections/{company_id}")
async def get_objection_responses(
    company_id: str,
    objection_type: Optional[str] = Query(None, description="Filter nach Einwand-Typ"),
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Holt Einwand-Antworten für eine Firma.
    
    Kann auch ohne aktive Session verwendet werden.
    """
    service = LiveAssistService(db)
    return service.get_objection_responses(
        company_id=company_id,
        objection_type=objection_type,
    )


@router.get("/objections")
async def get_all_objection_responses(
    objection_type: Optional[str] = Query(None, description="Filter nach Einwand-Typ"),
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Holt Einwand-Antworten (ohne Company-Filter).
    """
    service = LiveAssistService(db)
    return service.get_objection_responses(
        objection_type=objection_type,
    )


@router.get("/knowledge/{vertical}")
async def get_vertical_knowledge(
    vertical: str,
    knowledge_type: Optional[str] = Query(None, description="Filter nach Wissens-Typ"),
    query: Optional[str] = Query(None, description="Suchbegriff"),
    limit: int = Query(10, ge=1, le=50),
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Holt Branchenwissen für ein Vertical.
    """
    service = LiveAssistService(db)
    return service.get_vertical_knowledge(
        vertical=vertical,
        knowledge_type=knowledge_type,
        query=query,
        limit=limit,
    )


# =============================================================================
# FEEDBACK
# =============================================================================

@router.post("/query/{query_id}/feedback")
async def submit_query_feedback(
    query_id: str,
    request: QueryFeedbackRequest,
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Gibt Feedback zu einer Query-Antwort (v3.3 mit Learning).
    
    Unterstützt Intent/Objection-Korrekturen für Machine Learning.
    """
    service = LiveAssistServiceV3(db)
    success = service.submit_query_feedback(
        query_id=query_id,
        user_id=str(current_user.id),
        was_helpful=request.was_helpful,
        corrected_intent=request.corrected_intent,
        corrected_objection_type=request.corrected_objection_type,
        feedback_text=request.feedback_text
    )
    
    return {"success": success, "message": "Feedback gespeichert" if success else "Fehler beim Speichern"}


# =============================================================================
# COACH ANALYTICS (v3.3)
# =============================================================================

@router.get("/coach/insights", response_model=CoachInsightsResponse)
async def get_coach_insights(
    company_id: str = Query(..., description="Company ID"),
    days: int = Query(30, ge=1, le=90, description="Analyse-Zeitraum in Tagen"),
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Holt personalisierte Coach-Insights.
    
    Analysiert Mood-Patterns, Decision-Tendenzen und generiert
    personalisierte Tipps zur Verbesserung.
    """
    service = CoachAnalyticsService(db)
    insights = service.get_coach_insights(
        user_id=str(current_user.id),
        company_id=company_id,
        days=days
    )
    
    return insights.to_dict()


@router.get("/coach/performance", response_model=PerformanceMetricsResponse)
async def get_performance_metrics(
    company_id: str = Query(..., description="Company ID"),
    days: int = Query(30, ge=1, le=90, description="Analyse-Zeitraum in Tagen"),
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Holt Performance-Metriken für Live Assist.
    
    Zeigt Sessions, Response-Times, Outcomes etc.
    """
    service = CoachAnalyticsService(db)
    return service.get_performance_metrics(
        user_id=str(current_user.id),
        company_id=company_id,
        days=days
    )


@router.get("/coach/objections", response_model=List[ObjectionAnalyticsItem])
async def get_objection_analytics(
    company_id: str = Query(..., description="Company ID"),
    days: int = Query(30, ge=1, le=90, description="Analyse-Zeitraum in Tagen"),
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Holt Einwand-Analytics.
    
    Zeigt häufigste Einwände und Erfolgsraten der Antworten.
    """
    service = CoachAnalyticsService(db)
    return service.get_objection_analytics(
        company_id=company_id,
        days=days
    )


@router.post("/objection/{response_id}/used")
async def log_objection_response_used(
    response_id: str,
    was_successful: Optional[bool] = None,
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    """
    Loggt dass eine Einwand-Antwort verwendet wurde.
    
    Hilft bei der Optimierung der Erfolgsraten.
    """
    # Get current stats
    response = db.table("objection_responses").select(
        "times_used, success_rate"
    ).eq("id", response_id).single().execute()
    
    if not response.data:
        raise HTTPException(404, "Einwand-Antwort nicht gefunden")
    
    times_used = (response.data.get("times_used") or 0) + 1
    current_rate = response.data.get("success_rate")
    
    # Calculate new success rate if feedback provided
    new_rate = current_rate
    if was_successful is not None:
        if current_rate is None:
            new_rate = 1.0 if was_successful else 0.0
        else:
            # Weighted average
            success_count = current_rate * (times_used - 1)
            if was_successful:
                success_count += 1
            new_rate = success_count / times_used
    
    db.table("objection_responses").update({
        "times_used": times_used,
        "success_rate": new_rate,
    }).eq("id", response_id).execute()
    
    return {"success": True, "times_used": times_used, "success_rate": new_rate}


# =============================================================================
# WEBSOCKET (für Echtzeit Voice)
# =============================================================================

@router.websocket("/ws/{session_id}")
async def websocket_live_assist(
    websocket: WebSocket,
    session_id: str,
    db = Depends(get_db),
):
    """
    WebSocket für Echtzeit Voice-Assistenz.
    
    Client sendet Text (transkribiert von Voice), Server antwortet.
    
    Message Format (Client → Server):
    {
        "query": "Kunde sagt zu teuer",
        "type": "voice" | "text"
    }
    
    Message Format (Server → Client):
    {
        "response": "...",
        "response_short": "...",
        "intent": "objection",
        "objection_type": "price",
        "follow_up": "..."
    }
    """
    await websocket.accept()
    
    service = LiveAssistService(db)
    
    # TODO: Authenticate WebSocket connection
    # For now, we'll use a placeholder user_id
    
    try:
        while True:
            # Receive message
            data = await websocket.receive_json()
            
            query_text = data.get("query")
            query_type = data.get("type", "text")
            
            if not query_text:
                await websocket.send_json({
                    "error": "Query text required"
                })
                continue
            
            try:
                # Process query
                request = LiveQueryRequest(
                    session_id=session_id,
                    query_text=query_text,
                    query_type=query_type,
                )
                
                # TODO: Get user_id from WebSocket auth
                # For now, get from session
                session = db.table("live_assist_sessions").select(
                    "user_id"
                ).eq("id", session_id).single().execute()
                
                if not session.data:
                    await websocket.send_json({
                        "error": "Session not found"
                    })
                    continue
                
                user_id = session.data.get("user_id")
                
                response = service.process_query(user_id, request)
                
                # Send response
                await websocket.send_json({
                    "response": response.response_text,
                    "response_short": response.response_short,
                    "intent": response.detected_intent.value,
                    "objection_type": response.objection_type,
                    "follow_up": response.follow_up_question,
                    "technique": response.response_technique,
                    "response_time_ms": response.response_time_ms,
                })
                
            except Exception as e:
                await websocket.send_json({
                    "error": str(e)
                })
            
    except WebSocketDisconnect:
        # Client disconnected
        pass
    except Exception as e:
        print(f"WebSocket error: {e}")
        try:
            await websocket.close()
        except:
            pass


# =============================================================================
# 🔒 SECURITY MONITORING (Admin only)
# =============================================================================

@router.get(
    "/security/logs",
    response_model=dict,
    summary="Security-Logs abrufen (Admin only)",
    description="Gibt die letzten Security-Events zurück (Jailbreak-Versuche, sensible Anfragen)."
)
async def get_security_logs(
    limit: int = Query(50, ge=1, le=100, description="Maximale Anzahl an Logs"),
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user)
):
    """
    🔒 Gibt Security-Logs zurück.
    
    Nur für Admins sichtbar.
    """
    # TODO: Admin-Check einbauen
    # if not current_user.is_admin:
    #     raise HTTPException(status_code=403, detail="Admin access required")
    
    from ...config.prompts.locked_block import get_security_log, get_security_stats
    
    return {
        "logs": get_security_log(limit),
        "stats": get_security_stats(),
    }


@router.get(
    "/security/stats",
    response_model=dict,
    summary="Security-Statistiken abrufen",
    description="Gibt Statistiken über Security-Events zurück."
)
async def get_security_statistics(
    db = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user)
):
    """
    📊 Gibt Security-Statistiken zurück.
    """
    from ...config.prompts.locked_block import get_security_stats
    
    return get_security_stats()

File: salesflow-app/src/api/liveAssist.ts
/**
 * ╔════════════════════════════════════════════════════════════════════════════╗
 * ║  LIVE ASSIST API                                                           ║
 * ║  API Functions für Echtzeit-Verkaufsassistenz                              ║
 * ╚════════════════════════════════════════════════════════════════════════════╝
 */

import { API_CONFIG } from '../services/apiConfig';
import { supabase } from '../services/supabase';

// Get base URL from config
const API_BASE_URL = API_CONFIG.baseUrl;
import type {
  StartSessionRequest,
  StartSessionResponse,
  LiveQueryRequest,
  LiveQueryResponse,
  EndSessionRequest,
  SessionStatsResponse,
  QuickFactItem,
  ObjectionResponseItem,
  VerticalKnowledgeItem,
  ObjectionType,
  FactType,
} from '../types/liveAssist';
import type {
  CoachInsightsResponse,
  PerformanceMetrics,
  ObjectionAnalyticsItem,
} from '../types/coachInsights';

// =============================================================================
// HELPER
// =============================================================================

async function getAuthHeaders(): Promise<Record<string, string>> {
  try {
    const { data: { session } } = await supabase.auth.getSession();
    
    if (!session?.access_token) {
      // Ohne Token nur Content-Type Header
      return {
        'Content-Type': 'application/json',
      };
    }
    
    return {
      'Authorization': `Bearer ${session.access_token}`,
      'Content-Type': 'application/json',
    };
  } catch {
    return {
      'Content-Type': 'application/json',
    };
  }
}

async function apiRequest<T>(
  endpoint: string,
  options: RequestInit = {},
): Promise<T> {
  const headers = await getAuthHeaders();
  
  const response = await fetch(`${API_BASE_URL}${endpoint}`, {
    ...options,
    headers: {
      ...headers,
      ...options.headers,
    },
  });
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Unbekannter Fehler' }));
    throw new Error(error.detail || `HTTP ${response.status}`);
  }
  
  return response.json();
}

// =============================================================================
// SESSION MANAGEMENT
// =============================================================================

/**
 * Startet eine Live Assist Session.
 * Trigger: User sagt "Bin mit Kunde" oder aktiviert manuell.
 */
export async function startSession(
  request: StartSessionRequest,
): Promise<StartSessionResponse> {
  return apiRequest<StartSessionResponse>('/live-assist/start', {
    method: 'POST',
    body: JSON.stringify(request),
  });
}

/**
 * Beendet eine Live Assist Session.
 */
export async function endSession(
  request: EndSessionRequest,
): Promise<{ success: boolean; message: string }> {
  return apiRequest('/live-assist/end', {
    method: 'POST',
    body: JSON.stringify(request),
  });
}

/**
 * Holt Session-Statistiken.
 */
export async function getSessionStats(
  sessionId: string,
): Promise<SessionStatsResponse> {
  return apiRequest<SessionStatsResponse>(`/live-assist/session/${sessionId}`);
}

// =============================================================================
// QUERY PROCESSING
// =============================================================================

/**
 * Sendet eine Live-Anfrage.
 * Schnelle Antwort - optimiert für Echtzeit.
 */
export async function sendQuery(
  request: LiveQueryRequest,
): Promise<LiveQueryResponse> {
  return apiRequest<LiveQueryResponse>('/live-assist/query', {
    method: 'POST',
    body: JSON.stringify(request),
  });
}

// =============================================================================
// QUICK ACCESS (ohne Session)
// =============================================================================

/**
 * Holt Quick Facts für eine Firma.
 */
export async function getQuickFacts(options: {
  companyId?: string;
  vertical?: string;
  factType?: FactType;
  keyOnly?: boolean;
  limit?: number;
}): Promise<QuickFactItem[]> {
  const params = new URLSearchParams();
  
  if (options.factType) params.append('fact_type', options.factType);
  if (options.keyOnly) params.append('key_only', 'true');
  if (options.limit) params.append('limit', options.limit.toString());
  if (options.vertical) params.append('vertical', options.vertical);
  
  const queryString = params.toString();
  const endpoint = options.companyId 
    ? `/live-assist/facts/${options.companyId}${queryString ? `?${queryString}` : ''}`
    : `/live-assist/facts${queryString ? `?${queryString}` : ''}`;
  
  return apiRequest<QuickFactItem[]>(endpoint);
}

/**
 * Holt Einwand-Antworten.
 */
export async function getObjectionResponses(options: {
  companyId?: string;
  objectionType?: ObjectionType;
}): Promise<ObjectionResponseItem[]> {
  const params = new URLSearchParams();
  
  if (options.objectionType) params.append('objection_type', options.objectionType);
  
  const queryString = params.toString();
  const endpoint = options.companyId
    ? `/live-assist/objections/${options.companyId}${queryString ? `?${queryString}` : ''}`
    : `/live-assist/objections${queryString ? `?${queryString}` : ''}`;
  
  return apiRequest<ObjectionResponseItem[]>(endpoint);
}

/**
 * Holt Branchenwissen.
 */
export async function getVerticalKnowledge(options: {
  vertical: string;
  knowledgeType?: string;
  query?: string;
  limit?: number;
}): Promise<VerticalKnowledgeItem[]> {
  const params = new URLSearchParams();
  
  if (options.knowledgeType) params.append('knowledge_type', options.knowledgeType);
  if (options.query) params.append('query', options.query);
  if (options.limit) params.append('limit', options.limit.toString());
  
  const queryString = params.toString();
  const endpoint = `/live-assist/knowledge/${options.vertical}${queryString ? `?${queryString}` : ''}`;
  
  return apiRequest<VerticalKnowledgeItem[]>(endpoint);
}

// =============================================================================
// FEEDBACK
// =============================================================================

/**
 * Gibt Feedback zu einer Query-Antwort.
 */
export async function submitQueryFeedback(
  queryId: string,
  wasHelpful: boolean,
): Promise<{ success: boolean; message: string }> {
  return apiRequest(`/live-assist/query/${queryId}/feedback`, {
    method: 'POST',
    body: JSON.stringify({ was_helpful: wasHelpful }),
  });
}

/**
 * Loggt dass eine Einwand-Antwort verwendet wurde.
 */
export async function logObjectionResponseUsed(
  responseId: string,
  wasSuccessful?: boolean,
): Promise<{ success: boolean; times_used: number; success_rate: number }> {
  const params = wasSuccessful !== undefined 
    ? `?was_successful=${wasSuccessful}` 
    : '';
  
  return apiRequest(`/live-assist/objection/${responseId}/used${params}`, {
    method: 'POST',
  });
}

// =============================================================================
// COACH ANALYTICS
// =============================================================================

/**
 * Holt personalisierte Coach-Insights.
 * Analysiert Mood und Decision Patterns über einen Zeitraum.
 */
export async function getCoachInsights(
  companyId: string,
  days: number = 30,
): Promise<CoachInsightsResponse> {
  const params = new URLSearchParams();
  params.append('days', days.toString());
  
  return apiRequest<CoachInsightsResponse>(
    `/live-assist/coach/insights?company_id=${companyId}&${params.toString()}`
  );
}

/**
 * Holt Performance-Metriken (Sessions, Response Times, Outcomes).
 */
export async function getPerformanceMetrics(
  companyId: string,
  days: number = 30,
): Promise<PerformanceMetrics> {
  return apiRequest<PerformanceMetrics>(
    `/live-assist/coach/performance?company_id=${companyId}&days=${days}`
  );
}

/**
 * Holt Einwand-Analytics (welche Einwände, wie erfolgreich).
 */
export async function getObjectionAnalytics(
  companyId: string,
  days: number = 30,
): Promise<ObjectionAnalyticsItem[]> {
  return apiRequest<ObjectionAnalyticsItem[]>(
    `/live-assist/coach/objection-analytics?company_id=${companyId}&days=${days}`
  );
}

// =============================================================================
// WEBSOCKET
// =============================================================================

/**
 * Erstellt eine WebSocket-Verbindung für Echtzeit-Assistenz.
 */
export function createLiveAssistWebSocket(
  sessionId: string,
  onMessage: (data: {
    response: string;
    response_short?: string;
    intent: string;
    objection_type?: string;
    follow_up?: string;
    technique?: string;
    response_time_ms: number;
    error?: string;
  }) => void,
  onError?: (error: Event) => void,
  onClose?: () => void,
): WebSocket {
  const wsUrl = API_BASE_URL.replace('http', 'ws');
  const ws = new WebSocket(`${wsUrl}/live-assist/ws/${sessionId}`);
  
  ws.onmessage = (event) => {
    try {
      const data = JSON.parse(event.data);
      onMessage(data);
    } catch (e) {
      console.error('WebSocket parse error:', e);
    }
  };
  
  ws.onerror = (event) => {
    console.error('WebSocket error:', event);
    onError?.(event);
  };
  
  ws.onclose = () => {
    onClose?.();
  };
  
  return ws;
}

/**
 * Sendet eine Nachricht über WebSocket.
 */
export function sendWebSocketQuery(
  ws: WebSocket,
  query: string,
  type: 'text' | 'voice' = 'text',
): void {
  if (ws.readyState === WebSocket.OPEN) {
    ws.send(JSON.stringify({ query, type }));
  } else {
    console.error('WebSocket not open');
  }
}

// =============================================================================
// NAMED EXPORT OBJECT
// =============================================================================

export const liveAssistApi = {
  // Session
  startSession,
  endSession,
  getSessionStats,
  
  // Query
  sendQuery,
  
  // Quick Access
  getQuickFacts,
  getObjectionResponses,
  getVerticalKnowledge,
  
  // Feedback
  submitQueryFeedback,
  logObjectionResponseUsed,
  
  // Coach Analytics
  getCoachInsights,
  getPerformanceMetrics,
  getObjectionAnalytics,
  
  // WebSocket
  createLiveAssistWebSocket,
  sendWebSocketQuery,
};

